rfcs/ion_templates.md

```diff
@@ -0,0 +1,472 @@
+
+
+## Invoking a template
+
+### Invocation syntax
+
+```js
+// Invoke template 0 with the following parameters
+@0("Toyota", "Camry", 2017)
+```
+
+Parameters to a template invocation can themselves be template invocations. There is no restriction on which templates can be invoked as parameters.
+
+### Expanded data
+
+```js
+{ 
+  make: "Toyota",
+  model: "Camry",
+  year: 2017,
+  frame: "sedan",
+  numberOfWheels: 4,
+  transmission: "automatic",
+  hasAirbags: true
+}
+```
+
+In binary Ion, the template invocation requires 18 bytes to encode while the expanded data would require 42 bytes.
+
+## Binary layout
+
+See the [Ion 1.0 binary encoding spec](http://amzn.github.io/ion-docs/docs/binary.html#typed-value-formats) for a detailed explanation of the terms used in the diagram below.
+
+```
+            7       4 3       0
+           +---------+---------+
+Template   |   15    |    L    |
+           +---------+---------+======+
+           :     length [VarUInt]     :
+           +--------------------------+
+           |     template ID [VarUInt]|
+           +--------------------------+
+           :           value          :
+           +==========================+
+                         ⋮
+```
+
+* `L` values of `0` and `null` are available to be used as sentinels. 
+* This document treats `L`=0 as a template blank: `@_`
+* `L`=14 is a variable `length` encoding.
+* `L`=15 is reserved for future specs to accommodate not-yet-designed template invocation methods. (e.g. “splicing”)
+* Templates have a minimum length of 2 (1 byte type descriptor, 1 byte template ID)
+* A template blank (`L`=0, `@_`) found outside of a template definition or template invocation is illegal and should be treated as an error.
+* Template invocations with an ID that cannot be resolved are illegal and should be treated as an error. 
+    * (This differs from `symbol`s with an ID that cannot be resolved. `symbol`s are values while `template`s are an encoding detail. If a `Reader` encounters an unresolvable `symbol`, it can still tell the user that it is positioned over a symbol and even provide the unresolvable ID number.)
+
+
+
+## Text encoding within s-expressions
+
+Because `@` is a valid operator character within an s-expression, template blanks in the text format must be annotated with the special `$ion_template` annotation. Per the spec, symbols starting with `$ion` are reserved and may not be used in user data.
+
+### Blanks in a template definition s-expression
+
+```js
+$ion_template_table::{
+  templates : [{ // template 0
+    ($ion_template::@_)  
+  }]
+}  
+```
+
+### Template invocation inside an s-expression
+
+```js
+// Invoke template 0 from within an s-expression.
+($ion_template::@0(123))
+// ((123))
+```
+
+
+
+## Restrictions
+
+* Ion templates perform simple expansions. They involve no branching, looping, or recursion. They are guaranteed to terminate.
+* A template always expands to a single, complete Ion value.
```

Comment:
Aha. From this then I assume that when a template is invoked and the number of parameters passed is less than the number of `@_` in the template's definition is thus an error. Is that correct? 

This still does not answer the case of passing more parameters than needed.